src/main/scala/akka/persistence/jdbc/query/scaladsl/JdbcReadJournal.scala

```diff
@@ -135,39 +134,63 @@ class JdbcReadJournal(config: Config)(implicit val system: ExtendedActorSystem)
+  private def currentJournalEventsByTag(tag: String, offset: Long, max: Long, latestOrdering: MaxOrderingId): Source[EventEnvelope, NotUsed] = {
+    if (latestOrdering.maxOrdering < offset) Source.empty
+    else {
+      readJournalDao.eventsByTag(tag, offset, latestOrdering.maxOrdering, max)
+        .mapAsync(1)(Future.fromTry)
+        .mapConcat { case (repr, _, row) =>
+          adaptEvents(repr).map(r => EventEnvelope(Sequence(row.ordering), r.persistenceId, r.sequenceNr, r.payload))
+        }
+    }
+  }
+
+  /**
+    * @param terminateAfterOffset If None, the stream never completes. If a Some, then the stream will complete once a
+    *                             query has been executed which might return an event with this offset (or a higher offset).
+    *                             The stream may include offsets higher than the value in terminateAfterOffset, since the last batch
+    *                             will be returned completely.
+    */
+  private def eventsByTag(tag: String, offset: Long, terminateAfterOffset: Option[Long]): Source[EventEnvelope, NotUsed] = {
     import akka.pattern.ask
-    implicit val askTimeout = Timeout(readJournalConfig.journalSequenceRetrievalConfiguration.askTimeout)
-    Source.fromFuture(orderingActor.ask(GetMaxOrderingId).mapTo[MaxOrderingId])
-      .flatMapConcat { latestOrdering =>
-        if (latestOrdering.maxOrdering < offset) Source.empty
-        else readJournalDao.eventsByTag(tag, offset, latestOrdering.maxOrdering, max)
-      }
-      .mapAsync(1)(Future.fromTry)
+    implicit val askTimeout: Timeout = Timeout(readJournalConfig.journalSequenceRetrievalConfiguration.askTimeout)
+    /* We unfold with 3 parameters:
+     * - The first parameter (from) determines the starting offset (i.e. the minimum value) to query from.
+     * - The second parameter (completeImmediately) signifies that we have queried events until at least
+     *   `terminateAfterOffset`, therefore the stream no longer needs to continue.
+     * - The third parameter (retrieveImmediately) is a boolean which determines whether the database query should be
+     *   executed immediately, or after a delay.
+     */
+    Source.unfoldAsync[(Long, Boolean, Boolean), Seq[EventEnvelope]]((offset, false, false)) {
+      case (from: Long, completeImmediately: Boolean, retrieveImmediately: Boolean) =>
+        def retrieveNextBatch(): Future[Some[((Long, Boolean, Boolean), Seq[EventEnvelope])]] = {
+          for {
+            queryUntil <- journalSequenceActor.ask(GetMaxOrderingId).mapTo[MaxOrderingId]
+            xs <- currentJournalEventsByTag(tag, from, readJournalConfig.maxBufferSize, queryUntil).runWith(Sink.seq)
+          } yield {
+            val isFullBatch = xs.size == readJournalConfig.maxBufferSize
+            val completeStreamAfterThisBatch = !isFullBatch && terminateAfterOffset.exists(_ <= queryUntil.maxOrdering)
```

Comment:
I don't think it's need to check if batch is full. If the target offset is already reached we are safe to say that we don't need another iteration.

Reply:
We do need this check.

Say we query the journal sequence actor, it states that the maxOrdering id is 1000. We then execute with the following parameters:
```
bufferSize = 100
queryFrom = 50 // user provided offset
queryUntil = 1000 // max offset as retrieved from journalSequenceActor
```
We receive events 51 to 151. The pitfall here is that `queryUntil` is a max used in the query, but that does not mean that we actually receive all events until queryUntil

Reply:
Yes, you are right. I didn't consider that case.